### PR TITLE
Config with pure python

### PR DESCRIPTION
**Pull request by jugmac00**

### CODE REVIEW

File: .meta.cfg [hunk 1]
```diff
@@ -0,0 +1,7 @@
+# Generated from:
+# https://github.com/zopefoundation/meta/tree/master/config/pure-python
+[meta]
+template = pure-python
+commit-id = 05894360cac01f1027889dd14a23ae372e808b3f
+fail-under = 0
```
Comment (sallner) [inline .meta.cfg:6]: Could you please set the value, that was previously set in tox.ini (86) if that is still valid?

File: src/zc/zodbrecipes/tests.py [hunk 2]
```diff
@@ -80,5 +74,6 @@ def test_suite():
 
         ))
 
+
 if __name__ == '__main__':
```
Comment (icemac) [inline src/zc/zodbrecipes/tests.py:78]: This line and the following is no longer needed.

File: src/zc/zodbrecipes/tests.py [hunk 3]
```diff
@@ -15,15 +15,10 @@
...
-import os
-import pkg_resources
 import re
-import shutil
-import sys
-import tempfile
 import unittest
 import zc.buildout.testing
-import zope.testing
+import zope.testing  # noqa: F401
```
Comment (sallner) [inline src/zc/zodbrecipes/tests.py:21]: Why do we need this import here?

File: tox.ini [hunk 4]
```diff
@@ -1,30 +1,60 @@
...
+deps =
+    flake8
+    check-manifest
+    check-python-versions
 commands =
-    coverage run -m zope.testrunner --test-path=src []
-    coverage report --fail-under=86
+    flake8 src setup.py
+    check-manifest
+    check-python-versions .
```
Comment (icemac) [inline tox.ini:33]: You seem not to use the latest version of the script, I eliminated the dot at the end of is line as it is not necessary.

File: tox.ini [hunk 5]
```diff
@@ -1,30 +1,60 @@
...
+[testenv:coverage]
+basepython = python3
 deps =
     coverage
+    coverage-python-version
+    zope.testrunner
+commands =
+    coverage run -m zope.testrunner --test-path=src []
+    coverage html
+    coverage report -m --fail-under=0
```
Comment (icemac) [inline tox.ini:44]: This had been `--fail-under=86`, please change `.meta.cfg` accordingly so this number is used here.

File: tox.ini (hunk 4 above)
Comment (jugmac00) [reply tox.ini:33]: Actually, I applied an earlier version, and now I applied the current version, and tox.ini does not get updated. So I fixed it manually.

Comment (jugmac00) [reply tox.ini:33]: Ok, the pull failed because of the changed `packages.txt`...

File: tox.ini (hunk 5 above)
Comment (jugmac00) [reply tox.ini]: Now, this is surprising and should be mentioned clearly in the documentation.

When I run coverage on the master (without my changes), it fails with 39% - but there not only this package gets tested, but all in site-packages.

I'll have a look how to fix this.

Comment (jugmac00) [reply tox.ini]: When I ran coverage on the current master, as mentioned above, it covers the wrong stuff and even then, this package only has very low test coverage.

```
.tox/coverage/lib/python3.7/site-packages/zope/testrunner/find.py                             231     74    68%
.tox/coverage/lib/python3.7/site-packages/zope/testrunner/formatter.py                        610    405    34%
.tox/coverage/lib/python3.7/site-packages/zope/testrunner/garbagecollection.py                 37     15    59%
.tox/coverage/lib/python3.7/site-packages/zope/testrunner/interfaces.py                        27      0   100%
.tox/coverage/lib/python3.7/site-packages/zope/testrunner/layer.py                             10      3    70%
.tox/coverage/lib/python3.7/site-packages/zope/testrunner/listing.py                           13      5    62%
.tox/coverage/lib/python3.7/site-packages/zope/testrunner/logsupport.py                        20      4    80%
.tox/coverage/lib/python3.7/site-packages/zope/testrunner/options.py                          187     70    63%
.tox/coverage/lib/python3.7/site-packages/zope/testrunner/process.py                           22     12    45%
.tox/coverage/lib/python3.7/site-packages/zope/testrunner/profiling.py                         49     28    43%
.tox/coverage/lib/python3.7/site-packages/zope/testrunner/refcount.py                          61     51    16%
.tox/coverage/lib/python3.7/site-packages/zope/testrunner/runner.py                           668    324    51%
.tox/coverage/lib/python3.7/site-packages/zope/testrunner/selftest.py                           8      0   100%
.tox/coverage/lib/python3.7/site-packages/zope/testrunner/shuffle.py                           23      9    61%
.tox/coverage/lib/python3.7/site-packages/zope/testrunner/statistics.py                        21      2    90%
.tox/coverage/lib/python3.7/site-packages/zope/testrunner/tb_format.py                         52     31    40%
src/zc/zodbrecipes/__init__.py                                                                152    135    11%
src/zc/zodbrecipes/tests.py                                                                    40      3    92%
---------------------------------------------------------------------------------------------------------------
TOTAL                                                                                       22120  13488    39%
ERROR: InvocationError for command /home/jugmac00/Repos/tests/zc.zodbrecipes/.tox/coverage/bin/coverage report --fail-under=86 (exited with code 2)
_________________________________________________________________ summary _________________________________________________________________
ERROR:   coverage: commands failed
```

Comment (jugmac00) [reply tox.ini]: So, no - I won't set coverage to 86.

Comment (jugmac00) [reply tox.ini]: When you have a look at the build history on travis, there was no lint nor coverage runner.

https://travis-ci.org/github/zopefoundation/zc.zodbrecipes/builds

File: .meta.cfg (hunk 1 above)
Comment (jugmac00) [reply .meta.cfg]: No, because this was never run before, and test coverage is way below that number.

File: src/zc/zodbrecipes/tests.py (hunk 3 above)
Comment (jugmac00) [reply src/zc/zodbrecipes/tests.py:21]: For me it looks like it gets used in line 49
```
    zc.buildout.testing.install('zope.testing', test)
```
but who knows? This doctest thingy is still a mysery.

File: tox.ini (hunk 5 above)
Comment (mgedmin) [reply tox.ini]: Buildout recipes need special magic to measure coverage correctly.  See https://github.com/zopefoundation/z3c.recipe.tag/pull/7

This is documented in [buildout's testing.txt](https://github.com/buildout/buildout/blob/0185dc6a87f417160378547c776fb6da0c016e88/src/zc/buildout/testing.txt#L123-L152).

File: src/zc/zodbrecipes/tests.py (hunk 3 above)
Comment (mgedmin) [reply src/zc/zodbrecipes/tests.py:21]: No, that line easy_installs zope.testing into the temporary buildout working tree.

I think the unused zope.testing import can be safely removed.

File: setup.py [hunk 6]
```diff
@@ -16,8 +18,6 @@ def read(*rnames):
     '**********************\n'
     )
 
-open('doc.txt', 'w').write(long_description)
-
```
Comment (mgedmin) [inline setup.py:20]: thank you thank you thank YOU!

File: setup.py [hunk 7]
```diff
@@ -30,34 +30,34 @@ def read(*rnames):
...
-    author_email = "jim@zope.com",
-    description = "ZC Buildout recipes for ZODB",
-    license = "ZPL 2.1",
-    keywords = "zodb buildout",
+    author="Jim Fulton",
+    author_email="jim@zope.com",
+    description="ZC Buildout recipes for ZODB",
+    license="ZPL 2.1",
+    keywords="zodb buildout",
     url='http://svn.zope.org/'+name,
```
Comment (mgedmin) [inline setup.py:40]: Um please make this point to https://github.com/zopefoundation/zc.zodbrecipes.

File: src/zc/zodbrecipes/__init__.py [hunk 8]
```diff
@@ -247,7 +248,7 @@ def install(self):
             else:
                 try:
                     port = int(address)
-                except:
+                except:  # noqa: E722
```
Comment (mgedmin) [inline src/zc/zodbrecipes/__init__.py:251]:
```suggestion
                except ValueError:
```

File: tox.ini [hunk 9]
```diff
@@ -1,30 +1,62 @@
...
+[testenv:coverage]
+basepython = python3
 deps =
     coverage
+    coverage-python-version
+    zope.testrunner
+commands =
+    coverage run -m zope.testrunner --test-path=src []
+    coverage html
+    coverage report -m --fail-under=18
```
Comment (mgedmin) [inline tox.ini:44]: Measuring buildout recipe coverage need special magic, documented in https://github.com/buildout/buildout/blob/0185dc6a87f417160378547c776fb6da0c016e88/src/zc/buildout/testing.txt#L123-L152

File: src/zc/zodbrecipes/tests.py (hunk 3 above)
Comment (jugmac00) [reply src/zc/zodbrecipes/tests.py]: Thank you!

File: setup.py (hunk 6 above)
Comment (jugmac00) [reply setup.py:20]: You are welcome.

File: src/zc/zodbrecipes/__init__.py (hunk 8 above)
Comment (jugmac00) [reply src/zc/zodbrecipes/__init__.py]: So, I trust you scanned the context... as `int(None)` would throw a `TypeError` ...

File: tox.ini (hunk 5 above)
Comment (jugmac00) [reply tox.ini]: Thanks a lot!

Now, let's see how the config tool likes these changes....

File: tox.ini (hunk 9 above)
Comment (jugmac00) [reply tox.ini]: Magic applied!

**Comment by jugmac00** [comment]
Huh, coverage works on my machine, but fails on Travis. 

Any hints for me?

Thank you.

File: tox.ini (hunk 5 above)
Comment (jugmac00) [reply tox.ini]: Ofc - it does not like the changes... and overwrites them...

File: tox.ini [hunk 10]
```diff
@@ -1,30 +1,66 @@
...
 deps =
     coverage
+    coverage-python-version
+    zope.testrunner
+setenv =
+    COVERAGE_PROCESS_START={toxinidir}/tox.ini
+commands =
+    coverage run -m zope.testrunner --test-path=src []
+    coverage html
+    coverage combine
```
Comment (mgedmin) [inline tox.ini:46]: You need to run `coverage combine` before `coverage html`.
```suggestion
    coverage combine
    coverage html
```

Comment (jugmac00) [reply tox.ini]: Awesome! At some time I have to read the manual carefully... same goes for the other 125 tools I use...

Comment (mgedmin) [reply tox.ini]: PR review is a perfectly cromulent way of knowledge transfer.

(It's why I comment on already merged PRs that are perfectly fine, but maybe show that the author didn't know of a particular nifty feature they could've used.)

**Comment by jugmac00** [comment]
Now, this was a difficult birth.

Thanks a lot @mgedmin @icemac and @sallner

**Pull request merged**